### ISSUE TITLE
DPI

### ISSUE THREAD

**Issue by jimmyff** [DavBfr/dart_pdf#139]
I'm trying to generate 300dpi poker card sized files however I can't find a way to change the DPI. The constants in 'PdfPageFormat' suggest maybe the library only supports the default 72dpi?

I've done the conversion dpi conversion and provided those values as my width & height to 'PdfPageFormat' but then printers think that my output documents are larger than A4 rather than the size of a poker card.

Does the library support setting a higher DPI? If not is it something you'd consider?

Thanks

**Comment by DavBfr** [comment]
If your poker card is designed with text and vector graphics, then there is no DPI involved. Only your printer will raster to its best possible resolution.
If you use bitmaps, you're in charge to provide the right image DPI. The library will not resize the bitmaps. If you want a 300 DPI bitmap, create an image with 300 pixels width for each paper inch.

**Comment by jimmyff** [comment]
Ah okay, apologies -i'm not terribly clued up on printing stuff - i was just going by the mac print preview. So the PDF's generated via this shouldn't be a problem if I send the files to printing companies?

One little concern is that the PDFs generated seems to crash my mac's preview app, although it does open correctly in Chrome with a minor glitch (it shows a single grey border below the icon images in the top left - screenshot included). 

In Affinity illustrator the pdf looks perfectly as expected. Although I need to specify the DPI and color palette (I wonder if there is a way I could specify this in the PDFs so it knows?). 

Thanks

Example PDF:
[card-01.pdf](https://github.com/DavBfr/dart_pdf/files/3566583/card-01.pdf)

The border glitch showing on chrome:
<img width="376" alt="Screenshot 2019-09-02 at 16 23 23" src="https://user-images.githubusercontent.com/298850/64124236-5e2c5000-cd9e-11e9-8a6c-86cacf2e8694.png">

**Comment by DavBfr** [comment]
Maybe you forgot some await in your code and an image is not fully loaded and creates a non-conformant PDF file.

**Comment by jimmyff** [comment]
After quite a bit of testing, it's seems to be the images I'm using. The images that i have cropped cause the issue. Ones that I haven't cropped work fine. I've tried cropping the images in both pixelmator and affinity photo and both have the same issue. I've also tried messing with their color palettes etc.  Do you know of any way I can debug this further?

These are the images that cause mac preview to break:
![icon_core](https://user-images.githubusercontent.com/298850/64127827-4ad3b180-cdab-11e9-8c5c-09fcdc51b06d.png)
![icon_hamstrings](https://user-images.githubusercontent.com/298850/64127828-4ad3b180-cdab-11e9-9664-5b2c9358d55d.png)
![icon_hips](https://user-images.githubusercontent.com/298850/64127830-4ad3b180-cdab-11e9-8634-1872067ed24d.png)
![icon_shoulders](https://user-images.githubusercontent.com/298850/64127832-4b6c4800-cdab-11e9-96b1-45a434ade571.png)

**Comment by DavBfr** [comment]
PNG images should not break anything. They are decoded as RGBA data and recompressed.
How do you load the images?

**Comment by jimmyff** [comment]
```dart
  final muscleIcons = {
    MuscleTypes.shoulders:
        imgsrc.decodeImage(File('assets/icon_shoulders.png').readAsBytesSync()),
    MuscleTypes.core:
        imgsrc.decodeImage(File('assets/icon_core.png').readAsBytesSync()),
    MuscleTypes.hips:
        imgsrc.decodeImage(File('assets/icon_hips.png').readAsBytesSync()),
    MuscleTypes.hamstrings:
        imgsrc.decodeImage(File('assets/icon_hamstrings.png').readAsBytesSync())
  };
  final icons = <Widget>[
    for (var muscleType in MuscleTypes.values)
      for (var i = 0, l = muscles[muscleType] ?? 0; i < l; i++)
        Container(
          width: iconSize,
          margin: EdgeInsets.only(bottom: marginSize / 2),
          child: Image(
              PdfImage(
                pdf.document,
                image: muscleIcons[muscleType].data.buffer.asUint8List(),
                width: muscleIcons[muscleType].width,
                height: muscleIcons[muscleType].height,
              ),
              fit: BoxFit.contain),
        )
  ];
```
These are then provided to a `Column` widget. I know this code works okay as if I change the file paths to one of my uncropped PNGs then the output PDFs open fine in the mac preview app.

**Comment by DavBfr** [comment]
I can't reproduce your issue. I used this code:
```dart
import 'dart:io';

import 'package:image/image.dart' as im;
import 'package:pdf/pdf.dart';
import 'package:pdf/widgets.dart';

void main() {
  final Document pdf = Document();
  Document.debug = true;

  final im1 = im.decodePng(File('im1.png').readAsBytesSync());
  final im2 = im.decodePng(File('im2.png').readAsBytesSync());
  final im3 = im.decodePng(File('im3.png').readAsBytesSync());
  print('${im2.width}x${im2.height}');
  print(im2.runtimeType);

  pdf.addPage(Page(build: (context) {
    return Center(
      child: Column(
        mainAxisSize: MainAxisSize.min,
        children: [
          for (var im in [im1, im2, im3])
            SizedBox(
              width: 200,
              height: 200,
              child: Image(
                PdfImage(
                  pdf.document,
                  image: im.getBytes(),
                  width: im.width,
                  height: im.height,
                ),
              ),
            ),
        ],
      ),
    );
  }));

  final File file = File('doc.pdf');
  file.writeAsBytesSync(pdf.save());
}
```

**Comment by jimmyff** [comment]
Your example worked for me too so I started adding code in until I found the issue again. It appears that this is the cause of the problem:

```dart
  final Document pdf = Document(
    compress: false,
  );
```

The `compress: false` seems to only causes an issue with PDF file using those cropped images - if I use the uncropped images then the file is okay. Maybe pixelmator / affinity designer is doing something to them (compressing?) that is causing issue with the PDF generation?

**Comment by DavBfr** [comment]
Ok, maybe an issue with the Ascii85 algorithm :(

**Comment by jimmyff** [comment]
Ah good luck with it, it's a great library. Sorry for the longwinded route to discovering the true cause of the issue.

**Comment by DavBfr** [comment]
Fixed in pdf 1.3.19
